compactor/periodic.go

```diff
@@ -61,50 +61,99 @@ func newPeriodic(clock clockwork.Clock, h time.Duration, rg RevGetter, c Compact
-			if t.clock.Now().Sub(initialWait) < t.period {
-				continue
-			}
+// periodically fetches revisions and ensures that
+// first element is always up-to-date for retention window
+func (t *Periodic) run() {
+	initialWait := t.clock.Now()
+	fetchInterval := t.getInterval()
+	retryInterval := t.getRetryInterval()
+
+	// e.g. period 9h with compaction period 1h, then retain up-to 9 revs
+	// e.g. period 12h with compaction period 1h, then retain up-to 12 revs
+	// e.g. period 20m with compaction period 20m, then retain up-to 1 rev
+	retentions := int(t.period / fetchInterval)
+	for {
+		t.revs = append(t.revs, t.rg.Rev())
+		if len(t.revs) > retentions {
+			t.revs = t.revs[1:]
+		}
 
-			rev, remaining := t.getRev()
-			if rev < 0 {
+		select {
+		case <-t.ctx.Done():
+			return
+		case <-t.clock.After(fetchInterval):
+			t.mu.RLock()
+			p := t.paused
+			t.mu.RUnlock()
+			if p {
 				continue
 			}
+		}
+
+		// no compaction until initial wait period
+		if t.clock.Now().Sub(initialWait) < t.period {
+			continue
+		}
+		rev := t.revs[0]
 
+		for i := 0; i < retryDivisor; i++ {
 			plog.Noticef("Starting auto-compaction at revision %d (retention: %v)", rev, t.period)
 			_, err := t.c.Compact(t.ctx, &pb.CompactionRequest{Revision: rev})
 			if err == nil || err == mvcc.ErrCompacted {
-				// move to next sliding window
-				t.revs = remaining
+				// compactor succeeds at revs[0], move sliding window
+				t.revs = t.revs[1:]
 				plog.Noticef("Finished auto-compaction at revision %d", rev)
-			} else {
-				plog.Noticef("Failed auto-compaction at revision %d (%v)", rev, err)
-				plog.Noticef("Retry after %v", interval)
+				break
+			}
+
+			// compactor fails at revs[0]:
+			//  1. retry revs[0], so long as revs[0] is up-to-date
+			//  2. retry revs[1], when revs[0] becomes stale
+			plog.Noticef("Failed auto-compaction at revision %d (%v)", rev, err)
+			plog.Noticef("Retry after %v", retryInterval)
+			paused := false
+			select {
+			case <-t.ctx.Done():
+				return
+			case <-t.clock.After(retryInterval):
+				t.mu.RLock()
+				paused = t.paused
+				t.mu.RUnlock()
+			}
+			if paused {
+				break
 			}
 		}
-	}()
+	}
+}
+
+// If given compaction period x is <1-hour, compact every x duration, with x retention window
+// (e.g. --auto-compaction-mode 'periodic' --auto-compaction-retention='10m', then compact every 10-minute).
+// If given compaction period x is >1-hour, compact every 1-hour, with x retention window
+// (e.g. --auto-compaction-mode 'periodic' --auto-compaction-retention='72h', then compact every 1-hour).
+func (t *Periodic) getInterval() time.Duration {
+	itv := t.period
+	if itv > time.Hour {
+		itv = time.Hour
+	}
+	return itv
+}
+
+const retryDivisor = 10
+
+// divide by 10 to retry faster
+// e.g. given period 2-hour, retry in 12-min rather than 1-hour (compaction period)
+func (t *Periodic) getRetryInterval() time.Duration {
+	itv := t.period
+	if itv > time.Hour {
+		itv /= retryDivisor
+	}
+	return itv
```

Comment:
When `t.period` is 10minutes for example, retryInterval will be 10 minutes. This is fine.
But `retryDivisor` is still 10 in this case, so `Run()` tries 10 times with 10 minute wait, meaning 100 minutes in total.
We may need to return retry divisor as well?

Reply:
Right. Just fixed. Think we've covered all corner cases now? :)
I will take another look with @fanminshi tomorrow.

Thanks @yudai